File: back-end/src/main/java/tn/ONT/gestion_ONT/service/SoldeCongeService.java
package tn.ONT.gestion_ONT.service;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

import org.springframework.stereotype.Service;

import lombok.RequiredArgsConstructor;
import tn.ONT.gestion_ONT.dto.Inputsoldedto;
import tn.ONT.gestion_ONT.dto.SoldeCongeDto;
import tn.ONT.gestion_ONT.entity.Employe;
import tn.ONT.gestion_ONT.entity.SoldeConge;
import tn.ONT.gestion_ONT.repository.EmployeRepository;
import tn.ONT.gestion_ONT.repository.SoldeCongeRepository;

@Service
@RequiredArgsConstructor
public class SoldeCongeService {

    private final SoldeCongeRepository soldeCongeRepository;
    private final EmployeRepository employeRepository;

    // ✅ Initialiser un solde pour un nouvel employé
    public SoldeCongeDto creerSolde(Inputsoldedto inputsoldedto) {
        Employe employe = employeRepository.findById(inputsoldedto.getEmployeId())
                .orElseThrow(() -> new RuntimeException("Employé introuvable"));
       Optional<SoldeConge>  soldeexiste = soldeCongeRepository.findByEmployeAndAnnee(employe,inputsoldedto.getAnnee());
               if(soldeexiste.isPresent())
               {
            	   soldeexiste.get().setSoldeDisponible(inputsoldedto.getSoldeInitial());
            	   soldeexiste.get().setSoldePris(0);
            	   return SoldeCongeDto.fromEntity(soldeCongeRepository.save(soldeexiste.get()));
            	   
               }
        SoldeConge solde = SoldeConge.builder()
                .annee(inputsoldedto.getAnnee())
                .soldeDisponible(inputsoldedto.getSoldeInitial())
                .soldePris(0)
                .employe(employe)
                .build();

        return SoldeCongeDto.fromEntity(soldeCongeRepository.save(solde));
    }

    // ✅ Consulter le solde d'un employé pour une année donnée
    public SoldeCongeDto consulterSolde(Long employeId, int annee) {
        SoldeConge solde = soldeCongeRepository.findByEmployeAndAnnee(
                employeRepository.findById(employeId).orElseThrow(), annee)
                .orElseThrow(() -> new RuntimeException("Solde non trouvé"));

        return SoldeCongeDto.fromEntity(solde);
    }

    // ✅ Historique des soldes
    public List<SoldeCongeDto> historiqueSoldes(Long employeId) {
        return soldeCongeRepository.findAllByEmployeId(employeId).stream()
                .map(SoldeCongeDto::fromEntity)
                .collect(Collectors.toList());
    }

    // ✅ Mettre à jour manuellement un solde (admin/RH)
    public SoldeCongeDto modifierSolde(Long soldeId, int nouveauSoldeDisponible) {
        SoldeConge solde = soldeCongeRepository.findById(soldeId)
                .orElseThrow(() -> new RuntimeException("Solde introuvable"));

        solde.setSoldeDisponible(nouveauSoldeDisponible);
        return SoldeCongeDto.fromEntity(soldeCongeRepository.save(solde));
    }
}

File: front-end/src/app/services-gestion-ont/solde-conge.service.ts
import { Injectable } from '@angular/core';
import { environment } from '../../environment/environment';
import { HttpClient, HttpParams } from '@angular/common/http';
import { SoldeConge } from '../models-gestion-ont/demande-conge';
import { Observable } from 'rxjs';
import { Inputsoldedto } from '../models-gestion-ont/inputsolde';
import { SoldeCongeDto } from '../models-gestion-ont/SoldeCongeDto';

@Injectable({
  providedIn: 'root'
})
export class SoldeCongeService {
  
  baseUrl = environment.API_BASE_URL + "/solde-conge";
  
  constructor(private http: HttpClient) {}
  
  // Créer un solde pour un employé
creerSoldee(input: Inputsoldedto): Observable<SoldeCongeDto> {
    return this.http.post<SoldeCongeDto>(`${this.baseUrl}/Initialiser-solde`, input);
  }
  
  // Consulter le solde d'un employé pour une année donnée
  consulterSolde(employeId: number, annee: number): Observable<SoldeConge> {
    return this.http.get<SoldeConge>(`${this.baseUrl}/${employeId}/${annee}`);
  }
  
  // Obtenir l'historique des soldes d'un employé
  historiqueSoldes(employeId: number): Observable<SoldeConge[]> {
    return this.http.get<SoldeConge[]>(`${this.baseUrl}/historique/${employeId}`);
  }
  
  // Modifier manuellement un solde
  modifierSolde(soldeId: number, nouveauSoldeDisponible: number): Observable<SoldeConge> {
    return this.http.put<SoldeConge>(
      `${this.baseUrl}/${soldeId}`,
      null,
      { params: new HttpParams().set('nouveauSoldeDisponible', nouveauSoldeDisponible.toString()) }
    );
  }
}